### PR TITLE
Dijkstra algo

### PR DESCRIPTION
**Pull request by KaterynaMelnyk**
Added class WeightedGraph and the implementation of dijkstra algorithm

### CODE REVIEW

File: binarytree.cpp [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+#include <vector>
+#include <stack>
+#include <queue>
+
+namespace Trees
+{
+    void BinaryTree::new_node(int val)
+    {
+        Node *curr_node = root; // No need
```
Comment (SunshineJohn) [inline binarytree.cpp]: Be consistent with `Node *curr_node = root` and `Node* curr_node = root`.
Also `curr_node` and `currNode`

File: dijkstra.h [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+            minElem = vectorOfDistances[i];
+            indexMinElement = i;
+        }
+    }
+    return indexMinElement;
+}
+
+template <typename TYPE_GRAPH, typename TYPE_WEIGHT>
+std::vector<TYPE_WEIGHT> DijkstraAlgo(TYPE_GRAPH graph, size_t sourceNode) {
+
```
Comment (SunshineJohn) [inline dijkstra.h]: Remove empty line.

File: dijkstra.h [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+    size_t numberOfNodes = graph.getNumberOfNodes();
+    std::vector<bool> sptSet;
+    std::vector<TYPE_WEIGHT> distances;
+
```
Comment (SunshineJohn) [inline dijkstra.h]: Remove empty line.

File: dijkstra.h [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+        }
+    }
+    return indexMinElement;
+}
+
+template <typename TYPE_GRAPH, typename TYPE_WEIGHT>
+std::vector<TYPE_WEIGHT> DijkstraAlgo(TYPE_GRAPH graph, size_t sourceNode) {
+
+    size_t numberOfNodes = graph.getNumberOfNodes();
+    std::vector<bool> sptSet;
```
Comment (SunshineJohn) [inline dijkstra.h]: Reserve memory in prior.

File: dijkstra.h [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+    while( i < numberOfNodes) {
+        size_t indexCurrentNode;
+
+        indexCurrentNode = minDistance(distances, sptSet);
+        sptSet[indexCurrentNode] = 1;
+
+        //Change distance of all neighbors of CurrenNode
+        std::list<std::pair<size_t, TYPE_WEIGHT>> currentAdjList = graph.getAdjList(indexCurrentNode);
+
+        for (typename std::list<std::pair<size_t, TYPE_WEIGHT>>::iterator it = currentAdjList.begin(); it != currentAdjList.end(); ++it) {
```
Comment (SunshineJohn) [inline dijkstra.h]: `for (auto it = currentAdjList.begin()...)`
or
`for (auto pair : currentAdjList)`

File: graph.h [hunk 6]
```diff
@@ -0,0 +1,102 @@
...
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+#include <list>
+#include <utility>
+
+namespace Graph {
+
+    class Graph {
+
```
Comment (SunshineJohn) [inline graph.h]: Remove the empty line.

File: graph.h [hunk 7]
```diff
@@ -0,0 +1,102 @@
...
+    template <typename TYPE_WEIGHT>
+    void WeightedGraph<TYPE_WEIGHT>::addEdge(size_t sourceNode, size_t destinationNode, TYPE_WEIGHT weight) {
+        adjList[sourceNode].push_back(std::make_pair(destinationNode, weight));
+        adjList[destinationNode].push_back(std::make_pair(sourceNode, weight));
+    }
+
+    template <typename TYPE_WEIGHT>
+    void WeightedGraph<TYPE_WEIGHT>::printGraph() {
+        std::cout << "Weighted Graph: " << std::endl;
+        for (typename std::unordered_map<size_t, std::list<std::pair<size_t, TYPE_WEIGHT>>>::iterator  it = adjList.begin(); it != adjList.end(); ++it) {
```
Comment (SunshineJohn) [inline graph.h]: `for (auto it = adjList.begin().....)`
or
`for (auto part : adjList)`

File: graph.h [hunk 8]
```diff
@@ -0,0 +1,102 @@
...
+            std::cout << "Adj List of the node " << it->first << " is: ";
+            for (auto value: it->second) {
+                std::cout << "(" << value.first << ", " << value.second << ")" <<" ";
+            }
+            std::cout << std::endl;
+        }
+    }
+
+    template <typename TYPE_WEIGHT>
+    std::list<std::pair<size_t, TYPE_WEIGHT>> WeightedGraph<TYPE_WEIGHT>::getAdjList(size_t indexOfNode) {
```
Comment (SunshineJohn) [inline graph.h]: Do you plan to modify value? 
If not
`template <typename TYPE_WEIGHT> 
const std::list<std::pair<size_t, TYPE_WEIGHT>>& WeightedGraph<TYPE_WEIGHT>::getAdjList(size_t indexOfNode)`

File: main.cpp [hunk 9]
```diff
@@ -0,0 +1,93 @@
...
+    weightedGraphDijkstra.addEdge(7, 8, 7);
+    weightedGraphDijkstra.addEdge(6, 8, 6);
+    weightedGraphDijkstra.addEdge(2, 8, 2);
+    weightedGraphDijkstra.addEdge(2, 5, 4);
+    weightedGraphDijkstra.addEdge(3, 5, 14);
+
+    weightedGraphDijkstra.printGraph();
+
+    std::vector<int> distances;
+     distances = DijkstraAlgo<Graph::WeightedGraph<int>, int>(weightedGraphDijkstra, 0);
```
Comment (SunshineJohn) [inline main.cpp]: Unite 84 and 85 lines.

File: red_black_t.h [hunk 10]
```diff
@@ -1,34 +1,169 @@
...
-    NodeRb* get_root()
+    enum class NodeColor:char{red, black};
+
+    template <typename TYPE>
+    class Node_rb : public Node<TYPE>
+    {
+    public:
+        ~Node_rb()
+        {
+            delete parent;
```
Comment (SunshineJohn) [inline red_black_t.h]: Do not delete parent.

File: red_black_t.h [hunk 11]
```diff
@@ -1,34 +1,169 @@
...
+                        currNode->parent->parent->parent = currNode->parent;
+
+                        tmp_color = currNode->parent->color;
+                        currNode->parent->color = currNode->parent->parent->color;
+                        currNode->parent->parent->color = tmp_color;
+
+                    }
+                    else if (currNode->parent->data < currNode->parent->parent->data && currNode->data > currNode->parent->data)
+                    {
+
```
Comment (SunshineJohn) [inline red_black_t.h]: Empty braces.

File: red_black_t.h (hunk 10 above)
Comment (KaterynaMelnyk) [reply red_black_t.h]: Why not? This is a destructor

Comment (SunshineJohn) [reply red_black_t.h]: I already told you, each node removes its children and itself, that is why you don't need to remove parent.

File: binarytree.h [hunk 12]
```diff
@@ -1,291 +1,198 @@
...
             }
-            curr_node->right = nullptr;
+            currentNode->right = nullptr;
         }
         else
         {
-            Node<T> *min_node = curr_node->right;
-            Node<T> *p_node = curr_node;
+            Node<TYPE> *minNode = currentNode->right;
+            Node<TYPE> *pNode = currentNode;
```
Comment (SunshineJohn) [inline binarytree.h]: Name is not consistent.

File: dfs_bfs.h [hunk 13]
```diff
@@ -0,0 +1,106 @@
...
+            }
+        }
+
+        return dfsVector;
+    }
+
+    template <typename TYPE_TREE, typename TYPE>
+    std::vector<int> BFS(const TYPE_TREE& tree)
+    {
+        std::vector<int> bfsVector;
```
Comment (SunshineJohn) [inline dfs_bfs.h]: Still int?

File: dijkstra.h [hunk 14]
```diff
@@ -0,0 +1,56 @@
...
+        if (!sptSet[i] && vectorOfDistances[i] <= minElem) {
+            minElem = vectorOfDistances[i];
+            indexMinElement = i;
+        }
+    }
+    return indexMinElement;
+}
+
+template <typename TYPE_GRAPH, typename TYPE_WEIGHT>
+std::vector<TYPE_WEIGHT> DijkstraAlgo(TYPE_GRAPH graph, size_t sourceNode) {
```
Comment (SunshineJohn) [inline dijkstra.h]: You probably don't want to copy a graph, right?

File: graph.h [hunk 15]
```diff
@@ -0,0 +1,99 @@
+#ifndef GRAPH_H
+#define GRAPH_H
+
+#include <iostream>
+#include <unordered_map>
+#include <vector>
+#include <list>
+#include <utility>
+
+namespace Graph {
```
Comment (SunshineJohn) [inline graph.h]: `{ }` braces are not consistent through the code.

File: graph.h [hunk 16]
```diff
@@ -0,0 +1,99 @@
...
+        //TODO
+    }
+
+    template <typename TYPE_WEIGHT>
+    class WeightedGraph {
+    public:
+        WeightedGraph(size_t);
+
+        size_t getNumberOfNodes();
+        void addEdge(size_t sourceNode, size_t destinationNode, TYPE_WEIGHT weight);
```
Comment (SunshineJohn) [inline graph.h]: Is it source/dest node, or just index?

File: graph.h [hunk 17]
```diff
@@ -0,0 +1,99 @@
...
+    void getAdjMatrix(Graph* graph) {
+        //TODO
+    }
+
+    template <typename TYPE_WEIGHT>
+    class WeightedGraph {
+    public:
+        WeightedGraph(size_t);
+
+        size_t getNumberOfNodes();
```
Comment (SunshineJohn) [inline graph.h]: `size_t getNumberOfNodes() const;`

File: graph.h [hunk 18]
```diff
@@ -0,0 +1,99 @@
...
+        void addEdge(size_t sourceNode, size_t destinationNode, TYPE_WEIGHT weight);
+        void printGraph();
+
+        const std::list<std::pair<size_t, TYPE_WEIGHT>>& getAdjList(size_t indexOfNode);
```
Comment (SunshineJohn) [inline graph.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>& getAdjList(size_t indexOfNode) const;`

File: binarytree.h (hunk 12 above)
Comment (KaterynaMelnyk) [reply binarytree.h]: Everything is consistent.

File: dijkstra.h [hunk 19]
```diff
@@ -0,0 +1,56 @@
...
+    size_t numberOfNodes = graph.getNumberOfNodes();
```
Comment (SunshineJohn) [inline dijkstra.h]: `const size_t`

File: dijkstra.h [hunk 20]
```diff
@@ -0,0 +1,56 @@
...
+    std::vector<bool> sptSet(numberOfNodes);
```
Comment (SunshineJohn) [inline dijkstra.h]: `std::vector<bool> sptSet(numberOfNodes, 0);`
`std::vector<TYPE_WEIGHT> distances(numberOfNodes, INT_MAX);`

and remove loop belove.

File: dijkstra.h [hunk 21]
```diff
@@ -0,0 +1,56 @@
...
+
+    for (size_t i = 0; i < numberOfNodes; ++i) {
+        sptSet.push_back(0);
+        distances.push_back(INT_MAX);
+    }
+
+    distances[sourceNode] = 0;
+
+    size_t i = 0;
+    while( i < numberOfNodes) {
```
Comment (SunshineJohn) [inline dijkstra.h]: `while (i < numberOfNodes)`

File: dijkstra.h [hunk 22]
```diff
@@ -0,0 +1,56 @@
...
+        sptSet[indexCurrentNode] = 1;
+
+        std::list<std::pair<size_t, TYPE_WEIGHT>> currentAdjList = graph.getAdjList(indexCurrentNode);
+
+        for (auto pair : currentAdjList) {
+            if (!sptSet[pair->first] && distances[indexCurrentNode] + pair->second < distances[pair->first]) {
+                distances[pair->first] = distances[indexCurrentNode] + pair->second;
+            }
+        }
+        i++;
```
Comment (SunshineJohn) [inline dijkstra.h]: `++i`

File: graph.h [hunk 23]
```diff
@@ -0,0 +1,99 @@
...
+
+namespace Graph {
+    class Graph {
+    public:
+        Graph(size_t);
+
+        void addEdge(size_t sourceNode, size_t destinationNode);
+        void printGraph();
+
+        size_t getNumberOfNodes() {
```
Comment (SunshineJohn) [inline graph.h]: `size_t getNumberOfNodes() const`

File: graph.h [hunk 24]
```diff
@@ -0,0 +1,99 @@
...
+            return numberOfNodes;
+        }
+
+        Graph* getAdjMatrix(Graph* graph);
```
Comment (SunshineJohn) [inline graph.h]: What it should do?

File: binarytree.h [hunk 25]
```diff
@@ -1,291 +1,198 @@
 #ifndef BINARYTREE_H
 #define BINARYTREE_H
+
 #include <vector>
-#include <stack>
 #include <queue>
+#include <stack>
 
+#include "red_black_t.h"
```
[the remark this answers is not in the document]
Comment (KaterynaMelnyk) [reply binarytree.h]: Because I use class Node which is in red_black_t.h

[the remark this answers is not in the document]
Comment (SunshineJohn) [reply binarytree.h]: Then it's probably better to move Node class to separate file and not to have a binary tree dependent on red-black tree.

File: binarytree.h [hunk 26]
```diff
@@ -1,291 +1,169 @@
...
+    template <typename TYPE>
+    Node<TYPE>* BinaryTree<TYPE>::FindNode(TYPE newValueOfNode) {
+        Node<TYPE> *currentNode = root;
 
-        while (curr_node)
-        {
-            if (val == curr_node->data)
-            {
-                return curr_node;
+        assert(("There are no vertices in the tree.", currentNode));
```
Comment (SunshineJohn) [inline binarytree.h]: Assert is usually used to point on programmers' errors. Here it just indicates that the tree is empty.

File: binarytree.h [hunk 27]
```diff
@@ -1,291 +1,169 @@
...
             }
-            else
-            {
-                curr_node = curr_node->left;
+            else {
+                currentNode = currentNode->left;
             }
         }
-
+        std::cout << "There is no vertex with the value " << newValueOfNode << " in the tree." << std::endl;
```
Comment (SunshineJohn) [inline binarytree.h]: Are you sure that you'll always have a console to print on? It's better to handle this case in outer code (check if the result is nullptr and print in that case.)

File: binarytree.h [hunk 28]
```diff
@@ -1,291 +1,169 @@
...
-        Node<T> *parent_node = nullptr;
+    template <typename TYPE>
+    void BinaryTree<TYPE>::DeleteNode(TYPE newValueOfNode) {
+        Node<TYPE> *currentNode = root;
+        Node<TYPE> *parentNode = nullptr;
 
-        while (curr_node->data != val)
-        {
-            parent_node = curr_node;
+        assert(("There are no vertices in the tree to delete.", currentNode));
```
Comment (SunshineJohn) [inline binarytree.h]: Same as above.

File: binarytree.h [hunk 29]
```diff
@@ -1,291 +1,169 @@
...
-                curr_node = curr_node->left;
+            else if (newValueOfNode < currentNode->data) {
+                currentNode = currentNode->left;
             }
         }
 
-        if (val != curr_node->data)
-        {
+        if (newValueOfNode != currentNode->data) {
+            std::cout << "There is no vertex with the value " << newValueOfNode << " in the tree." << std::endl;
```
Comment (SunshineJohn) [inline binarytree.h]: Same as above.

File: dijkstra.h [hunk 30]
```diff
@@ -0,0 +1,53 @@
...
+    const size_t numberOfNodes = graph.GetNumberOfNodes();
+
+    std::vector<bool> checkedNodes(numberOfNodes, 0);
+    std::vector<TYPE_WEIGHT> distances(numberOfNodes, INT_MAX);
+
+    distances[indexSourceNode] = 0;
+
+    size_t i = 0;
+    while(i < numberOfNodes) {
+        size_t indexCurrentNode;
```
Comment (SunshineJohn) [inline dijkstra.h]: Merge with the line below.

File: dijkstra.h [hunk 31]
```diff
@@ -0,0 +1,53 @@
...
+std::vector<TYPE_WEIGHT> DijkstraAlgo(TYPE_GRAPH graph, size_t indexSourceNode) {
+    const size_t numberOfNodes = graph.GetNumberOfNodes();
+
+    std::vector<bool> checkedNodes(numberOfNodes, 0);
+    std::vector<TYPE_WEIGHT> distances(numberOfNodes, INT_MAX);
+
+    distances[indexSourceNode] = 0;
+
+    size_t i = 0;
+    while(i < numberOfNodes) {
```
Comment (SunshineJohn) [inline dijkstra.h]: `while (i < numberOfNodes)`

File: dijkstra.h [hunk 32]
```diff
@@ -0,0 +1,53 @@
...
+        size_t indexCurrentNode;
+
+        indexCurrentNode = MinDistance(distances, checkedNodes);
+        checkedNodes[indexCurrentNode] = 1;
+
+        std::list<std::pair<size_t, TYPE_WEIGHT>> currentAdjList = graph.GetAdjList(indexCurrentNode);
```
Comment (SunshineJohn) [inline dijkstra.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>&` I believe

File: graph.h [hunk 33]
```diff
@@ -0,0 +1,93 @@
...
+    template <typename TYPE_WEIGHT>
+    class WeightedGraph {
+    public:
+        WeightedGraph(size_t);
+
+        size_t GetNumberOfNodes() const;
+        void AddEdge(size_t sourceNode, size_t destinationNode, TYPE_WEIGHT weight);
+        void PrintGraph();
+
+        std::list<std::pair<size_t, TYPE_WEIGHT>>& GetAdjList(size_t indexOfNode);
```
Comment (SunshineJohn) [inline graph.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>& GetAdjList(size_t indexOfNode) const`

File: graph.h [hunk 34]
```diff
@@ -0,0 +1,93 @@
...
+#include <list>
+#include <utility>
+
+namespace Graph {
+    class Graph {
+    public:
+        Graph(size_t);
+
+        void AddEdge(size_t indexSourceNode, size_t indexDestinationNode);
+        void PrintGraph();
```
Comment (SunshineJohn) [inline graph.h]: `void PrintGraph() const;`

File: binarytree.h (hunk 26 above)
Comment (KaterynaMelnyk) [reply binarytree.h]: So does it mean that you never use assert ?

File: binarytree.h [hunk 35]
```diff
@@ -1,291 +1,175 @@
...
 
-    Node<T>* find_node(T val)
-    {
-        Node<T> *curr_node = root;
+    template <typename TYPE>
+    Node<TYPE>* BinaryTree<TYPE>::FindNode(TYPE newValueOfNode) {
+        Node<TYPE> *currentNode = root;
+
+        if (!currentNode) {
+            std::cout<< "There are no vertices in the tree to be found" <<std::endl;
```
Comment (SunshineJohn) [inline binarytree.h]: `.` in the end of the message.

File: dfs_bfs.h [hunk 36]
```diff
@@ -0,0 +1,146 @@
...
+        }
+        else if (currentNode->data > rightBorder) {
+            if (currentNode->left) {
+                CheckIfInRange(currentNode->left, leftBorder, rightBorder, nodeInList);
+            }
+        }
+    }
+
+    template<typename TYPE>
+    std::vector<int> SearchInRange(const BinaryTree<TYPE>& tree, int leftBorder, int rightBorder) {
```
Comment (SunshineJohn) [inline dfs_bfs.h]: Is it supposed to work only with `int`?

File: dfs_bfs.h [hunk 37]
```diff
@@ -0,0 +1,146 @@
...
+    std::vector<size_t> BFS(const Graph<TYPE_WEIGHT>& graph, size_t indexSourceNode){
+        std::vector<size_t> bfsVector;
+        std::queue<size_t> queueNodes;
+        std::vector<bool> checkedNodes(graph.GetNumberOfNodes(), false);
+        size_t currentNode = 0;
+
+        queueNodes.push(indexSourceNode);
+        checkedNodes[indexSourceNode] = true;
+
+        while(!queueNodes.empty()) {
```
Comment (SunshineJohn) [inline dfs_bfs.h]: `while (....)`

File: dijkstra.h [hunk 38]
```diff
@@ -0,0 +1,51 @@
...
+    std::vector<TYPE_WEIGHT> distances(numberOfNodes, INT_MAX);
+
+    distances[indexSourceNode] = 0;
+
+    size_t i = 0;
+    while(i < numberOfNodes) {
+        size_t indexCurrentNode = MinDistance(distances, checkedNodes);
+        checkedNodes[indexCurrentNode] = 1;
+
+        const std::list<std::pair<size_t, TYPE_WEIGHT>> currentAdjList = graph.GetAdjList(indexCurrentNode);
```
Comment (SunshineJohn) [inline dijkstra.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>&`

File: dfs_bfs.h [hunk 39]
```diff
@@ -0,0 +1,146 @@
...
+
+        stackNodes.push(indexSourceNode);
+        checkedNodes[indexSourceNode] = true;
+
+        while (!stackNodes.empty()) {
+            currentNode = stackNodes.top();
+            dfsVector.push_back(currentNode);
+            stackNodes.pop();
+
+            const std::list<std::pair<size_t, TYPE_WEIGHT>> neighborsOfCurrNode = graph.GetAdjList(currentNode);
```
Comment (SunshineJohn) [inline dfs_bfs.h]: `const std::list<std::pair<size_t, TYPE_WEIGHT>>&`

File: graph_kernels.h [hunk 40]
```diff
@@ -0,0 +1,17 @@
...
+
+#include <vector>
+#include "graph.h"
+
+namespace Graph {
+    template <typename TYPE_WEIGHT>
+    std::vector<std::vector<float>> WLSubtreeKrenel(std::vector<const Graph<TYPE_WEIGHT>>& vectorOfGraph) {
+        size_t numberOfGraphs = vectorOfGraph.size();
+        std::vector<std::vector<float>> kernelMatrix(numberOfGraphs, std::vector<float>(numberOfGraphs, 0));
+
```
Comment (SunshineJohn) [inline graph_kernels.h]: Removed these empty lines.

File: red_black_t.h [hunk 41]
```diff
@@ -1,34 +1,238 @@
...
+    };
+
+    template <typename TYPE>
+    RedBlackTree<TYPE>::~RedBlackTree() {
+        delete root;
+    }
+
+    template <typename TYPE>
+    void RedBlackTree<TYPE>::RotateLeftLeft(NodeRB<TYPE> *currentNode) {
+        NodeRB<TYPE> *rightCurrentChild = currentNode->parent->right;
```
Comment (SunshineJohn) [inline red_black_t.h]: Are you sure your `currentNode` is not `nullptr`? (check it in all methods)

File: dfs_bfs.h (hunk 37 above)
Comment (KaterynaMelnyk) [reply dfs_bfs.h]: What is wrong here?

File: dfs_bfs.h (hunk 39 above)
Comment (KaterynaMelnyk) [reply dfs_bfs.h]: Why ?

**Pull request merged**